Library/Homebrew/cmd/update.sh

```diff
@@ -293,14 +295,18 @@ EOS
   # ensure GIT_CONFIG is unset as we need to operate on .git/config
   unset GIT_CONFIG
 
+  # only allow one instance of brew update
+  source "$HOMEBREW_LIBRARY/Homebrew/utils/lock.sh"
```

Comment:
Probably the most important thing about this is that all the shell code that will be executed is loaded and parsed before we enter `homebrew-update`.